custom_components/tesla_custom/binary_sensor.py

```diff
@@ -1,40 +1,192 @@
+from teslajsonpy.car import TeslaCar
+from teslajsonpy.const import GRID_ACTIVE, RESOURCE_TYPE_BATTERY
+from teslajsonpy.energy import PowerwallSite
 
-from . import DOMAIN as TESLA_DOMAIN
-from .tesla_device import TeslaDevice
+from homeassistant.components.binary_sensor import (
+    BinarySensorDeviceClass,
+    BinarySensorEntity,
+)
+from homeassistant.core import HomeAssistant
 
+from . import TeslaDataUpdateCoordinator
+from .base import TeslaCarEntity, TeslaEnergyEntity
+from .const import DOMAIN
 
-async def async_setup_entry(hass, config_entry, async_add_entities):
-    """Set up the Tesla binary_sensors by config_entry."""
-    async_add_entities(
-        [
-            TeslaBinarySensor(
-                device,
-                hass.data[TESLA_DOMAIN][config_entry.entry_id]["coordinator"],
-            )
-            for device in hass.data[TESLA_DOMAIN][config_entry.entry_id]["devices"][
-                "binary_sensor"
-            ]
-        ],
-        True,
-    )
+_LOGGER = logging.getLogger(__name__)
 
 
-class TeslaBinarySensor(TeslaDevice, BinarySensorEntity):
-    """Implement an Tesla binary sensor for parking and charger."""
+async def async_setup_entry(hass: HomeAssistant, config_entry, async_add_entities):
+    """Set up the Tesla selects by config_entry."""
+    coordinator = hass.data[DOMAIN][config_entry.entry_id]["coordinator"]
+    cars = hass.data[DOMAIN][config_entry.entry_id]["cars"]
+    energysites = hass.data[DOMAIN][config_entry.entry_id]["energysites"]
+    entities = []
+
+    for car in cars.values():
+        entities.append(TeslaCarParkingBrake(hass, car, coordinator))
+        entities.append(TeslaCarOnline(hass, car, coordinator))
+        entities.append(TeslaCarChargerConnection(hass, car, coordinator))
+        entities.append(TeslaCarCharging(hass, car, coordinator))
+
+    for energysite in energysites.values():
+        if energysite.resource_type == RESOURCE_TYPE_BATTERY:
+            entities.append(TeslaEnergyBatteryCharging(hass, energysite, coordinator))
+            entities.append(TeslaEnergyGridStatus(hass, energysite, coordinator))
+
+    async_add_entities(entities, True)
+
+
+class TeslaCarParkingBrake(TeslaCarEntity, BinarySensorEntity):
+    """Representation of a Tesla car parking brake binary sensor."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        car: TeslaCar,
+        coordinator: TeslaDataUpdateCoordinator,
+    ) -> None:
+        """Initialize parking brake entity."""
+        super().__init__(hass, car, coordinator)
+        self.type = "parking brake"
+        self._attr_icon = "mdi:car-brake-parking"
+        self._attr_device_class = None
 
     @property
-    def device_class(self):
-        """Return the class of this binary sensor."""
+    def is_on(self):
+        """Return True if car shift state in park."""
+        return self._car.shift_state == "P"
```

Comment:
I noticed this binary sensor will report "Off" when the car is parked because `shift_state` is reported as `None`. 

I think a better naming convention would be a "shift state" sensor if there's a use case of needing to know what shift state the vehicle is in.

As for this parking brake sensor, it makes more sense to me to have "On" reported when the car is either in a shift state of "P" _or_ if the car is parked (reporting `None` from Tesla API) because physically, I'm pretty sure the parking brake is in fact engaged. Thoughts?

Reply:
That's fine. Just make sure you prepare an extensive breaking changes description since you're changing so much.

Reply:
Will do. I've been updating the first post as I'm making changes just so I don't forget anything. I'm just going to update this sensor to also report True when `shift_state` is `None` (e.g. car parked and off).